edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/planners/KubernetesPlanner.cs

```diff
@@ -0,0 +1,128 @@
+            ICombinedConfigProvider<T> combinedConfigProvider)
+        {
+            this.deviceNamespace = Preconditions.CheckNonWhiteSpace(deviceNamespace, nameof(deviceNamespace));
+            this.iotHubHostname = Preconditions.CheckNonWhiteSpace(iotHubHostname, nameof(iotHubHostname));
+            this.deviceId = Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId));
+            this.client = Preconditions.CheckNotNull(client, nameof(client));
+            this.commandFactory = Preconditions.CheckNotNull(commandFactory, nameof(commandFactory));
+            this.combinedConfigProvider = Preconditions.CheckNotNull(combinedConfigProvider, nameof(combinedConfigProvider));
+        }
+
+        public async Task<Plan> PlanAsync(
+            ModuleSet desired,
+            ModuleSet current,
+            IRuntimeInfo runtimeInfo,
+            IImmutableDictionary<string, IModuleIdentity> moduleIdentities)
+        {
+            Events.LogDesired(desired);
+            Events.LogCurrent(current);
+            Events.LogIdentities(moduleIdentities);
+
+            // Check that module names sanitize and remain unique.
+            var convertedNames = desired.Modules.Keys.Select(KubeUtils.SanitizeK8sValue).Distinct();
+            if (desired.Modules.Count != convertedNames.Count())
+            {
+                string nameList = desired.Modules.Keys.Join(",");
+                throw new InvalidIdentityException($"Deployment will cause a name collision in Kubernetes namespace, modules: [{nameList}]");
+            }
+
+            // TODO: improve this so it is generic for all potential module types.
+            if (!desired.Modules.Values.All(p => p is IModule<DockerConfig>))
+            {
+                throw new InvalidModuleException("Kubernetes deployment currently only handles type=DockerConfig");
+            }
+
+            Diff moduleDifference = desired.Diff(current);
+
+            Plan plan;
+            if (!moduleDifference.IsEmpty)
+            {
+                // The "Plan" here is very simple - if we have any change, publish all desired modules to a CRD.
+                // The CRD allows us to give the customer a Kubernetes-centric way to see the deployment
+                // and the status of that deployment through the "edgedeployments" API.
+                var k8sModules = desired.Modules.Select(m => new KubernetesModule<DockerConfig>(m.Value as IModule<DockerConfig>));
+
+                var crdCommand = new KubernetesCrdCommand<CombinedDockerConfig>(this.deviceNamespace, this.iotHubHostname, this.deviceId, this.client, k8sModules.ToArray(), Option.Some(runtimeInfo), this.combinedConfigProvider as ICombinedConfigProvider<CombinedDockerConfig>);
+                var planCommand = await this.commandFactory.WrapAsync(crdCommand);
+                var planList = new List<ICommand>
+                {
+                    planCommand
+                };
+                Events.PlanCreated(planList);
```

Comment:
Should events like this live closer to relevance? In this case, should the PlanCreated event happen inside of plan right before it returns? If the plan object fails, it was never created :)

Reply:
The plan itself is just populating data fields, so it's not significant behavior.   In terms of logging, you must CONFORM.
![OBEY CONFORM](https://media.giphy.com/media/CXcKi9NsRYa8U/giphy.gif)